js/ui/diagram/ui.diagram.js

```diff
@@ -49,11 +50,72 @@ class Diagram extends Widget {
         this._createComponent($options, DiagramOptions);
 
         !isServerSide && this._diagramInstance.createDocument($content[0]);
+        this._diagramInstance.onChanged = this._raiseDataChangeAction.bind(this);
     }
     _initDiagram() {
         const { DiagramControl } = getDiagram();
         this._diagramInstance = new DiagramControl();
     }
+
+    /**
+    * @name dxDiagramMethods.getData
+    * @publicName getData()
+    * @return string
+    */
+    getData() {
+        let value;
+        const { DiagramCommand } = getDiagram();
+        this._diagramInstance.commandManager.getCommand(DiagramCommand.Export).execute(function(data) { value = data; });
+        return value;
+    }
+    /**
+    * @name dxDiagramMethods.setData
+    * @publicName setData(value)
+    * @param1 data:string
+    */
+    setData(data) {
+        const { DiagramCommand } = getDiagram();
+        this._diagramInstance.commandManager.getCommand(DiagramCommand.Import).execute(data);
+        this._raiseDataChangeAction();
+    }
+
+    _getDefaultOptions() {
+        return extend(super._getDefaultOptions(), {
+            /**
+            * @name dxDiagramOptions.onDataChanged
+            * @extends Action
+            * @type function(e)
+            * @type_function_param1 e:object
+            * @type_function_param1_field4 data:string
+            * @action
+            */
+            onDataChanged: null,
+        });
+    }
+
+    _createDataChangeAction() {
+        this._dataChangeAction = this._createActionByOption("onDataChanged");
+    }
+    _raiseDataChangeAction() {
+        if(!this.option("onDataChanged")) return;
+
+        if(!this._dataChangeAction) {
+            this._createDataChangeAction();
+        }
+        this._dataChangeAction({
+            data: this.getData()
+        });
+    }
+
+    _optionChanged(args) {
+        switch(args.name) {
+            case "onDataChanged":
+                this._createDataChangeAction();
+                break;
+            default:
+                super._optionChanged(args);
```

Comment:
_invalidate?

Reply:
no need to invalidate if the event si assigned